ssl/ssl_transfer.cc

```diff
@@ -0,0 +1,674 @@
+  int present;
+  CBS value;
+  if (!CBS_get_optional_asn1_octet_string(cbs, &value, &present, tag)) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_INVALID_SSL3_STATE);
+    return false;
+  }
+  if (!present) {
+    return true;
+  }
+  size_t data_len = CBS_len(&value);
+  if (target_len > 0 && data_len != target_len) {
+    return false;
+  }
+  OPENSSL_memcpy(dst, CBS_data(&value), data_len);
+  return true;
+}
+
+// SSL3_STATE serialization.
+
+static const unsigned kS3Version = 1;
+
+static const unsigned kS3EstablishedSessionTag =
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 0;
+static const unsigned kS3SessionReusedTag =
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 1;
+static const unsigned kS3HostNameTag =
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 2;
+static const unsigned kS3ALPNSelectedTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 3;
+static const unsigned kS3NextProtoNegotiatedTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 4;
+static const unsigned kS3ChannelIdValidTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 5;
+static const unsigned kS3ChannelIdTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 6;
+static const unsigned kS3SendConnectionBindingTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 7;
+static const unsigned kS3PendingAppDataTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 8;
+static const unsigned kS3ReadBufferTag = 
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 9;
+
+// *** EXPERIMENTAL — DO NOT USE WITHOUT CHECKING ***
+// These SSL3_STATE serialization functions are developed to support SSL transfer.
+
+// ssl3_state_to_bytes serializes |in| to bytes stored in |cbb|.
+// It returns one on success and zero on failure.
+//
+// An SSL3_STATE is serialized as the following ASN.1 structure:
+//
+// SSL3State ::= SEQUENCE {
+//    version                           INTEGER (1),  -- SSL3_STATE structure version
+//    readSequence                      OCTET STRING,
+//    writeSequence                     OCTET STRING,
+//    serverRandom                      OCTET STRING,
+//    clientRandom                      OCTET STRING,
+//    sendAlert                         OCTET STRING,
+//    rwstate                           INTEGER,
+//    earlyDataReason                   INTEGER,
+//    previousClientFinished            OCTET STRING,
+//    previousClientFinishedLen         INTEGER,
+//    previousServerFinished            OCTET STRING,
+//    previousServerFinishedLen         INTEGER,
+//    emptyRecordCount                  INTEGER,
+//    warningAlertCount                 INTEGER,
+//    establishedSession                [0] SEQUENCE OPTIONAL,
+//    sessionReused                     [1] BOOLEAN OPTIONAL,
+//    hostName                          [2] OCTET STRING OPTIONAL,
+//    alpnSelected                      [3] OCTET STRING OPTIONAL,
+//    nextProtoNegotiated               [4] BOOLEAN OPTIONAL,
```

Comment:
//    nextProtoNegotiated               [4] OCTET STRING OPTIONAL,

Reply:
Will correct the comment.